design-system/src/theme-provider.tsx

```diff
@@ -1,94 +1,108 @@
     );
   }
-  return 'default';
-};
 
-const ThemeProvider = (props: ThemeProviderProps) => {
-  const root = useRef<HTMLElement>(document.querySelector(':root'));
-  const localScopeElement = useRef<HTMLDivElement>(null);
-  const [theme, setTheme] = useState<ThemeName>(validateTheme(props?.theme));
+  const vars = toVars(
+    themeOverrides && isObject(themeOverrides)
+      ? merge({}, themes[validTheme], themeOverrides)
+      : themes[validTheme]
+  );
 
-  const changeTheme = useCallback((newTheme: string) => {
-    setTheme(validateTheme(newTheme));
-  }, []);
+  Object.entries(vars).forEach(([key, value]) => {
+    target.style.setProperty(key, value);
+  });
+  target.setAttribute('data-theme', validTheme);
+};
 
+type ThemeProviderProps = {
+  parentSelector: typeof defaultParentSelector;
+  theme?: string;
+  themeOverrides?: Record<string, string>;
+};
+
+const ThemeProvider = (props: ThemeProviderProps) => {
+  const parentSelectorRef = useRef(props.parentSelector);
   useLayoutEffect(() => {
-    const vars = toVars(themes[theme]);
-    const targetElement =
-      props.scope === 'local' ? localScopeElement.current : root.current;
-    Object.entries(vars).forEach(([key, value]) => {
-      targetElement?.style.setProperty(key, value);
+    applyTheme({
+      newTheme: props.theme,
+      parentSelector: parentSelectorRef.current,
+      themeOverrides: props.themeOverrides,
     });
-  }, [theme, props.scope]);
-
-  const value = useMemo(() => {
-    return { theme, changeTheme };
-  }, [theme, changeTheme]);
+  }, [props.theme, props.themeOverrides]);
 
-  return (
-    <ThemeContext.Provider value={value}>
-      <div ref={localScopeElement}>{props.children}</div>
-    </ThemeContext.Provider>
-  );
+  return null;
+};
+ThemeProvider.defaultProps = {
+  parentSelector: defaultParentSelector,
 };
 
-const useTheme = () => {
-  const context = useContext(ThemeContext);
+const useTheme = (parentSelector = defaultParentSelector) => {
+  const [theme, setTheme] = useState<string>('default');
+  const parentSelectorRef = useRef(parentSelector);
 
-  warning(!isEmpty(context), `useTheme must be used within a ThemeProvider`);
+  // If we use 'useLayoutEffect' here, we would be trying to read the
+  // data attribute before it gets set from the effect in the ThemeProvider
+  useEffect(() => {
+    setTheme(parentSelectorRef.current()?.dataset.theme || 'default');
+  }, []);
 
-  return context;
+  // So consumers don't have to provide 'parentSelector' again as
+  // they already provided it in the hook call
+  /* const updateTheme = useRef(
+    ({ newTheme, themeOverrides }: Omit<TApplyTheme, 'parentSelector'>) => {
+      applyTheme({
+        newTheme,
+        parentSelector: parentSelectorRef.current,
+        themeOverrides,
+      });
+      setTheme(newTheme || 'default');
+    }
+  ); */
+  return useMemo(() => {
+    return { theme, applyTheme };
+  }, [theme]);
```

Comment:
@CarlosCortizasCT I had to revert to using the `applyTheme` where user needs to provide the `parentSelector` again. Unfortunately, with this implementation `parentSelector` function is stale, perhaps you could have a look into it if you find time (as I am not sure how to fix that). That can be now tested with rtl tests and in visual testing app -> `/theme-provider/interactive`

Reply:
The parent selector should be a function, not a ref, right? 🤔

Reply:
In this case we were using a ref to have a static reference to a function which shouldn't change. It is not a DOM reference.

I still don't get the use case of providing a different `parentSelector` for an already rendered hook, but I will take a deeper look and update accordingly.

Reply:
So I guess commenting previous suggested implementation it's due to tests using the API in a different way of what that implementation was expecting.

With the current implementation I see a "confusing" API in the hook.
I'll try to describe the current API
```
declare type TApplyTheme = {
    newTheme?: string;
    parentSelector?: () => HTMLElement | null;
    themeOverrides?: Record<string, string>;
};

declare const useTheme: (parentSelector?: () => HTMLElement | null) => {
    theme: string;
    applyTheme: ({ newTheme, parentSelector, themeOverrides, }: TApplyTheme) => void;
};
```
As you can see, when we call the hook in our component we have the chance to provide the `parentSelector` but we also have that chance when calling the `applyTheme` function of its result. So, if we use the `parentSelector` parameter in the `applyTheme`, why do we need the param in the hook itself for?

Since the hook and the provider are not connected by a React context, we need some way of linking them per instance. The provider will set the theme styles in a DOM element the hook will need to know about when changing.

The initial idea is that a hook will be linked to a specific provider by providing the same `parentSelector` the user used when rendering the provider. If we initialize the hook with that, there's no need to provide that callback in every `applyTheme` function call:
```
const getThemeRootElement = () => document.getElementById('theme-root');

const { theme, applyTheme } = useTheme(getThemeRootElement);

<button onClick={() => applyTheme({ newTheme: 'dark' })} />
```
In this case, the API will look like this:
```
declare type TApplyTheme = {
    newTheme?: string;
    parentSelector?: () => HTMLElement | null;
    themeOverrides?: Record<string, string>;
};

declare const useTheme: (parentSelector?: () => HTMLElement | null) => {
    theme: string;
    applyTheme: ({ newTheme, themeOverrides, }: TApplyTheme) => void;
};
```


On the other hand, we might want to go for another alternative and expect the `applyTheme` function called from users to always require the `parentSelector`. In that case, the `parentSelector` at hook initialization won't make sense:
```
const getThemeRootElement = () => document.getElementById('theme-root');

const { theme, applyTheme } = useTheme();

<button onClick={() => applyTheme({ newTheme: 'dark', parentSelector: getThemeRootElement })} />
```

In this case, the API will look like this:
```
declare type TApplyTheme = {
    newTheme?: string;
    parentSelector?: () => HTMLElement | null;
    themeOverrides?: Record<string, string>;
};

declare const useTheme: () => {
    theme: string;
    applyTheme: ({ newTheme, parentSelector, themeOverrides, }: TApplyTheme) => void;
};
```

Perhaps we can have a quick sync to explain this further.

Reply:
Thanks for your very detailed reply.
Having read it through, it seems to be the latter API makes more sense to me (which is closer to what we have had so far).
This would allow to call the generic `useTheme` hook e.g. only once, and use the exposed `applyTheme` function many times with various `parentSelector`s.

For me personally that makes more sense than the below:
```ts
// ... one useTheme hook call per theme
const { applyTheme: applyGlobalTheme } = useTheme(globalParentSelector);
const { applyTheme: applyLocalTheme } = useTheme(localParentSelector);
```

Reply:
> This would allow to call the generic useTheme hook e.g. only once, and use the exposed applyTheme function many times with various parentSelectors.

Bear in mind it's just the opposite. Every call to `useTheme` would bind to a specific `parentSelector` which should be the same provided to the `ThemeProvider` component you want this hook instance to control.
It's just what happens in your example.

You want both properties returned by the hook to always be bound to the same `parentSelector`.
Let's consider the previous test implementation and reading also the current theme:
```
const TestComponentWithThemeProvider = () => {
  const { theme, applyTheme } = useTheme();
  return (
    <>
      <button
        onClick={() => {
          applyTheme({
          applyGlobalTheme({
            themeOverrides: {
              colorSolid: 'red',
              colorSurface: 'yellow',
            },
            newTheme: 'dark',
            parentSelector: globalParentSelector,
          });
        }}
      >
        change global theme
      </button>
      <button
        onClick={() => {
          applyTheme({
          applyLocalTheme({
            themeOverrides: {
              colorSolid: 'green',
              colorSurface: 'tomato',
              customColor: '#BADA55',
            },
            newTheme: 'dark',
            parentSelector: localParentSelector,
          });
        }}
      >
...
```
In this scenario, we have only one hook instance so the same `theme` state is shared between changing local and global scoped themes, but in reality, we have two themes in the document.
That's why I think we need to have one hook instance per theme you want to control.

Thanks for updating the tests to comply with this suggestion 🙇